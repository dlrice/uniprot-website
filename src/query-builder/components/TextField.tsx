import React, { useEffect, useState } from 'react';

import { QueryBit, SearchTermType } from '../types/searchTypes';

const getStringValue = (value: string, prefix?: string) =>
  value.includes(' ')
    ? `"${prefix ? `${prefix}` : ''}${value?.trim()}"`
    : `${prefix ? `${prefix}` : ''}${value?.trim()}`;

const TextField: React.FC<{
  field: SearchTermType;
  type: string;
  initialValue?: string;
  handleChange: (queryBit: QueryBit) => void;
}> = ({ field, type, handleChange, initialValue = '' }) => {
  const [value, setValue] = useState(initialValue);

  useEffect(() => {
    if (value.length > 0) {
      if (field.term === 'All') {
        handleChange({
          [field.id]: value?.trim(),
        });
      } else if (
        field.valuePrefix &&
        field.term === 'database' &&
        value.trim() === '*'
      ) {
        // Query is faster with this hack
        handleChange({
          [field.id]: `(${field.term}:${field.valuePrefix})`,
        });
      } else {
        handleChange({
          [field.id]: `(${field.term}:${getStringValue(
            value,
            field.valuePrefix
          )})`,
        });
      }
    }
  }, [field, value, handleChange]);

  return (
    <label htmlFor={`input_${field.term}`}>
      {field.label}
      <input
        id={`input_${field.term}`}
        type={type}
        value={value}
        onChange={(e) => setValue(e.target.value)}
        placeholder={field.example}
      />
    </label>
  );
};

export default TextField;
