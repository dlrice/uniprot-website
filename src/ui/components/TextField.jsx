import React from 'react';

const dataTypes = { string: 'text', integer: 'number' };

const TextField = ({
  field, type, handleChange, value = '',
}) => (
  <div className="advanced-search__inputs" key={field.term}>
    <label htmlFor={`input_${field.term}`}>
      {field.label}
      <input
        id={`input_${field.term}`}
        type={dataTypes[type]}
        value={value}
        onChange={e => handleChange(e.target.value)}
        placeholder={field.example}
      />
    </label>
  </div>
);

export default TextField;
