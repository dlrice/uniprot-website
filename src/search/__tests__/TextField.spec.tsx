import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TextField from '../TextField';

describe('Range field', () => {
  const props = {
    field: {
      label: 'UniProtKB AC',
      itemType: 'single',
      term: 'accession',
      description: 'Search by UniProtKB Accession',
      example: 'P12345',
    },
    type: 'string',
    queryInput: {},
    handleChange: jest.fn(),
  };

  test('should render a text field', () => {
    const { asFragment, getByPlaceholderText } = render(
      <TextField {...props} />
    );
    const inputElt = getByPlaceholderText('P12345');
    expect(inputElt.value).toBe('');
    fireEvent.change(inputElt, { target: { value: 'Some term' } });
    expect(inputElt.value).toBe('Some term');
    expect(asFragment()).toMatchSnapshot();
  });
});
