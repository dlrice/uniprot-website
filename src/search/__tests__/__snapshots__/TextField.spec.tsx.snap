// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range field should render a text field 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextField
    field={
      Object {
        "description": "Search by UniProtKB Accession",
        "example": "P12345",
        "itemType": "single",
        "label": "UniProtKB AC",
        "term": "accession",
      }
    }
    handleChange={[MockFunction]}
    queryInput={Object {}}
    type="string"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "accession",
    "nodeType": "host",
    "props": Object {
      "children": <label
        htmlFor="input_accession"
      >
        UniProtKB AC
        <input
          id="input_accession"
          onChange={[Function]}
          placeholder="P12345"
          type="string"
          value=""
        />
      </label>,
      "className": "advanced-search__inputs",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "UniProtKB AC",
          <input
            id="input_accession"
            onChange={[Function]}
            placeholder="P12345"
            type="string"
            value=""
          />,
        ],
        "htmlFor": "input_accession",
      },
      "ref": null,
      "rendered": Array [
        "UniProtKB AC",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "input_accession",
            "onChange": [Function],
            "placeholder": "P12345",
            "type": "string",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "label",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "accession",
      "nodeType": "host",
      "props": Object {
        "children": <label
          htmlFor="input_accession"
        >
          UniProtKB AC
          <input
            id="input_accession"
            onChange={[Function]}
            placeholder="P12345"
            type="string"
            value=""
          />
        </label>,
        "className": "advanced-search__inputs",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "UniProtKB AC",
            <input
              id="input_accession"
              onChange={[Function]}
              placeholder="P12345"
              type="string"
              value=""
            />,
          ],
          "htmlFor": "input_accession",
        },
        "ref": null,
        "rendered": Array [
          "UniProtKB AC",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "input_accession",
              "onChange": [Function],
              "placeholder": "P12345",
              "type": "string",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "label",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
